tests/unit/test_charm.py

```diff
@@ -116,54 +102,48 @@ def test__alerting_config(self):
             self.harness.charm._alerting_config()
             self.assertEqual(sorted(logger.output), ["DEBUG:charm:No alertmanagers available"])
 
+    @patch("ops.model.Container.can_connect", MagicMock(return_value=False))
     @patch("charm.LokiOperatorCharm._loki_config")
-    def test__on_config_cannot_connect(self, mock_loki_config, *unused):
+    def test__on_config_cannot_connect(self, mock_loki_config):
         self.harness.set_leader(True)
         mock_loki_config.return_value = yaml.safe_load(LOKI_CONFIG)
-        self.harness.charm._container.can_connect = Mock()
-        self.harness.charm._container.can_connect.return_value = False
-        self.harness.update_config(yaml.safe_load(LOKI_CONFIG))
+
+        # Since harness was not started with begin_with_initial_hooks(), this must
+        # be emitted by hand to actually trigger _configure()
+        self.harness.charm.on.config_changed.emit()
         self.assertIsInstance(self.harness.charm.unit.status, WaitingStatus)
 
-    @patch("ops.testing._TestingPebbleClient.push")
+    @patch("ops.testing._TestingPebbleClient.push", Mock())
+    @patch("ops.model.Container.can_connect", MagicMock(return_value=True))
     @patch("charm.LokiOperatorCharm._loki_config")
-    def test__on_config_can_connect(self, mock_loki_config, *unused):
-        self.harness.set_leader(True)
+    def test__on_config_can_connect(self, mock_loki_config):
         mock_loki_config.return_value = yaml.safe_load(LOKI_CONFIG)
-        self.harness.charm._container.can_connect = Mock()
-        self.harness.charm._container.can_connect.return_value = True
-        self.harness.update_config(yaml.safe_load(LOKI_CONFIG))
-        self.assertIsInstance(self.harness.charm.unit.status, ActiveStatus)
+        self.harness.set_leader(True)
 
-    @patch("ops.testing._TestingPebbleClient.make_dir")
-    @patch("loki_server.LokiServer.version", new_callable=PropertyMock)
-    def test__provide_loki(self, mock_version, *unused):
-        mock_version.return_value = "3.14159"
+        # Since harness was not started with begin_with_initial_hooks(), this must
+        # be emitted by hand to actually trigger _configure()
```

Comment:
Is there a reason to keep `begin` and not replace it with `begin_with_initial_hooks`?

Reply:
For one, we need to wait for a [patch to the storage harness to merge](https://github.com/canonical/operator/pull/680) or the tests will blow up since the storage name uses hyphens.

The other is the [sheer amount of stuff happening in __init__](https://github.com/canonical/loki-k8s-operator/blob/829195d9db6e3b02687c170648afc16959992b8f/src/charm.py#L68). It would take a lot of mocking in the initial `setUp` to ensure that we could "pretend" that the k8s service was patched, that the Loki server was available (or mocked), etc. 

It's not impossible to do, just a lot of more changes in mocking instead of manually emitting one event here.

Reply:
Well said Ryan